tokio-util/src/sync/mpsc.rs

```diff
@@ -0,0 +1,145 @@
+//! Tokio-02 style MPSC channel.
+use std::{
+    future::Future,
+    marker::PhantomPinned,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tokio::sync::mpsc::{error::SendError, Permit};
+
+/// Fat sender with `poll_ready` support.
+#[derive(Debug)]
+// TODO: it doesn't seem we really need 'static here...
+pub struct Sender<T: 'static> {
+    // field order: `state` can contain reference to `inner`,
+    // so it must be dropped first.
+    /// State of this sender
+    state: State<T>,
+    /// Underlying channel
+    inner: tokio::sync::mpsc::Sender<T>,
+    /// Pinned marker
+    pinned: PhantomPinned,
+}
+
+impl<T: 'static> Sender<T> {
+    /// It is OK to get mutable reference to state, because it is not
+    /// self-referencing.
+    fn pin_project_state(self: Pin<&mut Self>) -> &mut State<T> {
+        unsafe { &mut Pin::into_inner_unchecked(self).state }
+    }
+
+    /// Sender must be pinned because state can contain references to it
+    fn pin_project_inner(self: Pin<&mut Self>) -> Pin<&mut tokio::sync::mpsc::Sender<T>> {
```

Comment:
This `Pin` doesn't do anything since `Sender` is `Unpin`. The compiler cannot help you with checking that it isn't moved.

Reply:
rustdoc says that Sender is `!Unpin`. I'll also add a test for it.

Reply:
`tokio::sync::mpsc::Sender` is `Unpin`.

Reply:
You're right.
I will remove Pin from return value and mark function as unsafe.